examples/adaptive_crag/application.py

```diff
@@ -0,0 +1,537 @@
+    """
+    query = state["query"]
+    try:
+        query = str(
+            ask_gemini.create(
+                messages=[user_message(exa_search_query_template(query=query))],  # type: ignore
+                response_model=ExaSearchKeywords,
+                max_retries=extraction_attempts,
+            )
+        )
+    except Exception as e:
+        print(f"Error in extract_keywords_for_web_search: {e}")
+    return {"exa_search_keywords": query}, state.update(exa_search_keywords=query)
+
+
+@action(reads=["exa_search_keywords", "docs_limit"], writes=["exa_search_results"])
+def search_exa(state: State) -> tuple[dict[str, list[str]], State]:
+    """
+    More details about the Exa API:
+        - https://docs.exa.ai/reference/getting-started
+        - https://github.com/exa-labs/exa-py
+    """
+    try:
+        exa_search_results = exa.search_and_contents(
+            query=state["exa_search_keywords"],
+            num_results=state["docs_limit"],
+            highlights=True,
+        )
+        exa_search_results = [res.highlights[0] for res in exa_search_results.results]
+    except Exception as e:
+        print(f"Error in search_exa: {e}")
+        exa_search_results = []
+    return {"exa_search_results": exa_search_results}, state.update(
+        exa_search_results=exa_search_results
+    )
+
+
+@action(
+    reads=["query", "lancedb_results", "exa_search_results", "chat_history"],
+    writes=["chat_history", "lancedb_results", "exa_search_results"],
+)
+def ask_assistant(
+    state: State, attempts: int = ATTEMPTS
+) -> tuple[dict[str, str], State]:
+    """
+    Combine the `chat_history`, `query`, `lancedb_results`, and `exa_search_results` to ask the assistant for an answer.
+    `chat_history`, `lancedb_results`, and `exa_search_results` are used as context for the assistant and can be empty lists.
+    """
+    query = state["query"]
+    lancedb_results = state.get("lancedb_results", [])
+    exa_search_results = state.get("exa_search_results", [])
+    messages = state["chat_history"]
+    context = lancedb_results + exa_search_results
+    if context:
+        context = "\n".join(context)
+        query = (
+            query.strip()
+            + "\n<additional_context>\n"
+            + context
+            + "\n</additional_context>"
+        )
+    messages.append({"role": "user", "content": query.strip()})
+    try:
+        response = ask_gemini.create(
+            messages=messages, response_model=str, max_retries=attempts
+        )
+    except Exception as e:
+        print(f"Error in ask_assistant: {e}")
+        response = "Sorry, please try again."
+    return (
+        {"assistant_response": response},  # type: ignore
+        state.append(chat_history=user_message(query))
+        .append(
+            chat_history=assistant_message(response)  # type: ignore
+        )
+        .update(lancedb_results=[], exa_search_results=[]),
```

Comment:
Why are the lance_db/exa results written as empty? Is this to allow a reset/refresh? Worth adding a comment.